src/main/java/org/apache/mesos/offer/constrain/AndRule.java

```diff
@@ -0,0 +1,88 @@
+package org.apache.mesos.offer.constrain;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.apache.mesos.Protos.Offer;
+import org.apache.mesos.Protos.Resource;
+import org.apache.mesos.Protos.TaskInfo;
+
+/**
+ * Wrapper for one or more another rules which returns the AND/intersection of those rules.
+ */
+public class AndRule implements PlacementRule {
+
+    private final Collection<PlacementRule> rules;
+
+    public AndRule(Collection<PlacementRule> rules) {
+        this.rules = rules;
+    }
+
+    public AndRule(PlacementRule... rules) {
+        this(Arrays.asList(rules));
+    }
+
+    @Override
+    public Offer filter(Offer offer) {
+        // Counts of how often each Resource passed each filter:
+        Map<Resource, Integer> resourceCounts = new HashMap<>();
+        for (PlacementRule rule : rules) {
```

Comment:
A matter of taste, but how do you feel about expressing this logic with streams and something like `Stream.allMatch`? It might expose the semantics more clearly.